app/javascript/src/ChangePassword/components/ChangePasswordHooks.jsx

```diff
@@ -1,87 +1,103 @@
+    },
+    length: {
+      minimum: 1,
+      message: '^isMandatory'
+    },
+    equality: {
+      attribute: PASSWORD,
+      message: '^mustMatch'
+    }
   }
 }
 
-const isFormValid = (formNode) => validateForm(formNode, validationConstraints)
-
-const getPasswordConfirmationError = (errors) => {
-  const passwordConfirmationError = (errors && errors[PASSWORD_CONFIRMATION]) && errors[PASSWORD_CONFIRMATION][0]
-  const passwordsMatchError = passwordConfirmationError && passwordConfirmationError.includes('is not equal to')
-  return {
-    passwordConfirmationError: !!passwordConfirmationError,
-    errorMessage: passwordsMatchError ? 'mustMatch' : 'isMandatory'
-  }
+const fieldsTemplate: FieldState = {
+  [PASSWORD]: '',
+  [PASSWORD_CONFIRMATION]: ''
 }
 
-const comparePasswordsLength = (event) => {
-  const passwordInput = event.currentTarget['user_password']
-  const passwordConfirmationInput = event.currentTarget['user_password_confirmation']
-  return passwordConfirmationInput.value.length >= passwordInput.value.length
+const fieldErrorsTemplate: FieldErrorsState = {
+  [PASSWORD]: [''],
+  [PASSWORD_CONFIRMATION]: ['']
 }
 
+const extractFirstErrorMessage = (errorMessageArray: string[]): string => errorMessageArray.slice(0, 1).pop()
+const isFieldValid = (errorMessageArray: string[]): boolean => !errorMessageArray.slice().pop().length
+
 const useFormState = () => {
+  const [fieldValues, setFieldValues] = useState(fieldsTemplate)
+  const [fieldErrors, setFieldErrors] = useState(fieldErrorsTemplate)
   const [isFormDisabled, setIsFormDisabled] = useState(true)
-  const [passwordValue, setPasswordValue] = useState('')
-  const [isPasswordValid, setIsPasswordValid] = useState(undefined)
-  const [passwordConfirmationValue, setPasswordConfirmationValue] = useState('')
-  const [isPasswordConfirmationValid, setIsPasswordConfirmationValid] = useState(undefined)
-  const [confirmationErrorMessage, setConfirmationErrorMessage] = useState('isMandatory')
-  const [validationStarted, setValidationStarted] = useState(false)
+  const [areFieldsPristine, setAreFieldsPristine] = useState({[PASSWORD]: true, [PASSWORD_CONFIRMATION]: true})
 
-  const onFormChange = (event) => {
-    const errors = isFormValid(event.currentTarget)
-    const { passwordConfirmationError, errorMessage } = getPasswordConfirmationError(errors)
-    const initValidation = comparePasswordsLength(event)
-    if (initValidation && !validationStarted) {
-      setValidationStarted(initValidation)
-    }
-    if (initValidation || validationStarted) {
-      setIsPasswordConfirmationValid(!passwordConfirmationError)
-    }
-    setConfirmationErrorMessage(errorMessage)
-    setIsFormDisabled(!!errors)
+  const onFormChange = (event: SyntheticInputEvent<HTMLInputElement>) => {
+    const validationErrors: ValidationErrors = validate(event.currentTarget, validationConstraints)
+    const visibleErrors = !!validationErrors && Object.keys(validationErrors)
```

Comment:
I'd watch out for the `!!` trick. The correct way would be `validationErrors !== undefined`.

Reply:
The "explicit" way you mean

Reply:
`validate` returns either an Object or `undefined`, the problem is that `!!{} === true` and we can't assure validate will never return that, either by error or by a change in their API.

So IMO here the verbose option is justified.

Reply:
That's why our type system is for, it return either `ValidationErrors` or `undefined`. If they change the API we will know and we are safe since we have our deps locked. If it fails because of that we can update our types and make the changes. 
The only way I'm making this change, if we update the types accordingly.